File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {    
    public bool IsSubtree(TreeNode root, TreeNode subRoot)
    {
        if (root == null)
            return false;

        return IsSameTree(root, subRoot)
            || IsSubtree(root.left, subRoot)
            || IsSubtree(root.right, subRoot);
    }

    private bool IsSameTree(TreeNode a, TreeNode b)
    {
        if (a == null || b == null)
            return a == b;

        return a.val == b.val
            && IsSameTree(a.left, b.left)
            && IsSameTree(a.right, b.right);
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        count = 0
        def traverse(node: Optional[TreeNode]) -> int:
            nonlocal count
            nonlocal k

            if not node:
                return None

            # go left
            value = traverse(node.left)

            if value is not None:
                return value

            # at root
            count += 1

            # check if we hit k
            if count == k:
                return node.val

            # go right
            value = traverse(node.right)

            if value is not None:
                return value

        return traverse(root)
